# BREATHING WAVE
## ANN using PyTorch
### 05 May 2023
***

## PART 1 : Data Preprocessing

### Importing Library

In [1]:
import pandas as pd
import numpy as np

### Import Dataset

In [2]:
df = pd.read_csv("dataset/breathing_waveform_data.csv").iloc[:, :-1] # get rid of last column ("notes")

### Filter the zeros values
> This will filtered the zeros value from all column (except first column)
>
> CAUSE : I think is natural for the first column to be 0.0 (because the time(X) still on 0 second)

In [3]:
zeros_val = df[df.iloc[:, 1:].eq(0).any(axis=1)]

In [4]:
zeros_val

0    1    2    3    4    5    6    7    8    9  ...   76   77   78  \
5473  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5474  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5475  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5476  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5477  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
6142  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6143  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6144  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6145  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6146  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

       79   80   81   82   83   84  labels  
5473  0.0  0.0  0.0  0.0  0.0  0.0  normal  
5474  0.0  0.0  0.0  0.0  0.0  0.0  normal  
5475  0.0  0.0  0.0  0.0  0.0  0.0  normal  
5476  0.0  0.0  0.0  0.0  0.0  0.0  normal  
5477  0.0  0.0  0.0  0.0  0.0  0.0  normal  
...   ...  ...  ...  ...  ...  ...     ...  
6142  0.0  0.0  0.0  0.0  0.0  0.0  normal  
6143  0.0  0.0  0.0  0.0  0.0  0.0  normal  
6144  0.0  0.0  0.0  0.0  0.0  0.0  normal  
6145  0.0  0.0  0.0  0.0  0.0  0.0  normal  
6146  0.0  0.0  0.0  0.0  0.0  0.0  normal  

[674 rows x 86 columns]

### Drop the table that has value zeros on it
> Size of original data : 26400 rows
>
> Size of zero values : 674 rows
>
> Result : 26400 - 674 = 25726 rows

In [5]:
df = df[~df.isin(zeros_val)].dropna()

In [6]:
df

0         1         2         3         4         5         6  \
0      0.483309  0.459790  0.431024  0.376565  0.295734  0.193290  0.066060   
1     -2.044518 -1.935588 -1.808629 -1.667919 -1.513497 -1.348760 -1.171044   
2     -1.213535 -1.269056 -1.323306 -1.375251 -1.430062 -1.485479 -1.529200   
3     -0.914806 -0.887726 -0.856065 -0.823527 -0.794551 -0.768074 -0.740895   
4     -1.547469 -1.458818 -1.362120 -1.264829 -1.164948 -1.060064 -0.954496   
...         ...       ...       ...       ...       ...       ...       ...   
26395 -0.152463 -0.164723 -0.165409 -0.152623 -0.118115 -0.066218 -0.010253   
26396 -0.164723 -0.165409 -0.152623 -0.118115 -0.066218 -0.010253  0.041637   
26397 -0.165409 -0.152623 -0.118115 -0.066218 -0.010253  0.041637  0.092217   
26398 -0.152623 -0.118115 -0.066218 -0.010253  0.041637  0.092217  0.140510   
26399 -0.118115 -0.066218 -0.010253  0.041637  0.092217  0.140510  0.188025   

              7         8         9  ...        76        77        78  \
0     -0.083445 -0.247221 -0.409374  ...  0.391514  0.452677  0.521407   
1     -0.972509 -0.759554 -0.547793  ...  0.138731 -0.053860 -0.241691   
2     -1.557172 -1.574662 -1.575457  ...  0.947940  0.996154  1.035743   
3     -0.713364 -0.685445 -0.652020  ... -0.478218 -0.571465 -0.684115   
4     -0.849448 -0.742812 -0.636614  ...  0.227050  0.130983  0.041438   
...         ...       ...       ...  ...       ...       ...       ...   
26395  0.041637  0.092217  0.140510  ... -0.336787 -0.306774 -0.280607   
26396  0.092217  0.140510  0.188025  ... -0.306774 -0.280607 -0.269843   
26397  0.140510  0.188025  0.240939  ... -0.280607 -0.269843 -0.260062   
26398  0.188025  0.240939  0.294399  ... -0.269843 -0.260062 -0.229981   
26399  0.240939  0.294399  0.340346  ... -0.260062 -0.229981 -0.167654   

             79        80        81        82        83        84  labels  
0      0.595845  0.661691  0.702932  0.708613  0.682564  0.637765    deep  
1     -0.417603 -0.582320 -0.738485 -0.889731 -1.037066 -1.174654    deep  
2      1.049543  1.024204  0.954716  0.844505  0.702445  0.541555    deep  
3     -0.817078 -0.966231 -1.122537 -1.264759 -1.376908 -1.461059    deep  
4     -0.038034 -0.106152 -0.163048 -0.210926 -0.253102 -0.290270    deep  
...         ...       ...       ...       ...       ...       ...     ...  
26395 -0.269843 -0.260062 -0.229981 -0.167654 -0.082300  0.004372   quick  
26396 -0.260062 -0.229981 -0.167654 -0.082300  0.004372  0.089958   quick  
26397 -0.229981 -0.167654 -0.082300  0.004372  0.089958  0.179209   quick  
26398 -0.167654 -0.082300  0.004372  0.089958  0.179209  0.264014   quick  
26399 -0.082300  0.004372  0.089958  0.179209  0.264014  0.343418   quick  

[25726 rows x 86 columns]

### Separate the X (data) and Y (label) & check for the null values

In [7]:
X = df.iloc[:, :-1]
Y = df.iloc[:, -1]

# Check if the data do not have any NULL 
print("X have a null? \t{}".format(X.isnull().values.any()))
print("Y have a null? \t{}".format(Y.isnull().values.any()))

X have a null? 	False
Y have a null? 	False


### Scalling the data (X) using StandardScaler

In [8]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

In [9]:
X

array([[ 1.57732521,  1.50096158,  1.40710808, ...,  2.29629039,
         2.21721275,  2.07692465],
       [-6.66012793, -6.30517583, -5.88996555, ..., -2.8575958 ,
        -3.34130049, -3.79623272],
       [-3.95219545, -4.13306073, -4.30872165, ...,  2.73447696,
         2.28147373,  1.76515392],
       ...,
       [-0.53665953, -0.49479087, -0.38205583, ...,  0.02545401,
         0.30168015,  0.59097159],
       [-0.49499252, -0.38233365, -0.21296997, ...,  0.30142842,
         0.5901747 ,  0.8657814 ],
       [-0.38253957, -0.21321113, -0.03062927, ...,  0.58922045,
         0.86429589,  1.12308954]])

### Transform the labe (Y) to one hot encoded

In [10]:
from sklearn.preprocessing import LabelEncoder
from keras.utils import np_utils

# encode class values as integers [0,0,0,0,0,0,0,1,1,1,1,1,2,2,2,2]
encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = encoder.transform(Y)

# convert integers to dummy variables (i.e. one hot encoded)
hot_y = np_utils.to_categorical(encoded_Y)

In [11]:
hot_y

array([[1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       ...,
       [0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 1.]], dtype=float32)

### Split the data (X) and label (Y) using train_test_split()
> I will set the split ratio 80% training and 20% testing

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, hot_y, test_size=.2, random_state=21)

## PART 2 : Building Neural Network using PyTorch

### Importing pytorch libraries

In [13]:
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F

### Convert all the training and testing data into PyTorch Tensor type variables

In [14]:
X_train = torch.Tensor(X_train)
X_test = torch.Tensor(X_test)
Y_train = torch.Tensor(Y_train)
Y_test = torch.Tensor(Y_test)

### Define the model acrhitecture

In [15]:
class MyModel(nn.Module):
    def __init__(self, input_size, output_size):
        super(MyModel, self).__init__()
        self.model = nn.Sequential(
            nn.Linear(input_size, 128),
            nn.ReLU(),
            nn.Linear(128, 256),
            nn.ReLU(),
            nn.Linear(256, output_size)
        )
    
    def forward(self, x):
        return self.model(x)

In [16]:
input_size = X_train.shape[1]
output_size = Y_train.shape[1]
model = MyModel(input_size, output_size)

### Initiliaze the loss function and optimizer

In [17]:
# Define the loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

### Training

In [18]:
num_of_epochs = 200
for epoch in range(num_of_epochs):
    # reset optimizer
    optimizer.zero_grad()
    
    # forward pass
    outputs = model(X_train)
    loss = criterion(outputs, Y_train)
    
    # backward pass and optimization
    loss.backward()
    optimizer.step()
    
    # Print the loss for monitoring the training progress
    print(f"Epoch {epoch+1}/{num_of_epochs}, Loss: {loss.item()}")
    

Epoch 1/200, Loss: 1.5349531173706055
Epoch 2/200, Loss: 1.5067249536514282
Epoch 3/200, Loss: 1.2525296211242676
Epoch 4/200, Loss: 1.2125089168548584
Epoch 5/200, Loss: 1.1518081426620483
Epoch 6/200, Loss: 1.0826128721237183
Epoch 7/200, Loss: 1.0155946016311646
Epoch 8/200, Loss: 0.9280393719673157
Epoch 9/200, Loss: 0.8735841512680054
Epoch 10/200, Loss: 0.8301234841346741
Epoch 11/200, Loss: 0.80362468957901
Epoch 12/200, Loss: 0.8266608119010925
Epoch 13/200, Loss: 0.7976644039154053
Epoch 14/200, Loss: 0.7406864762306213
Epoch 15/200, Loss: 0.7912884950637817
Epoch 16/200, Loss: 0.7277735471725464
Epoch 17/200, Loss: 0.7363976240158081
Epoch 18/200, Loss: 0.7170231342315674
Epoch 19/200, Loss: 0.715427577495575
Epoch 20/200, Loss: 0.6949171423912048
Epoch 21/200, Loss: 0.6888406276702881
Epoch 22/200, Loss: 0.6837934255599976
Epoch 23/200, Loss: 0.6703577637672424
Epoch 24/200, Loss: 0.6869359612464905
Epoch 25/200, Loss: 0.6608166098594666
Epoch 26/200, Loss: 0.667327582836151

## PART 3 : Evaluate the model

### Counting the average loss from testing data

In [19]:
# set the model to evaluation mode (e.g disable operation like dropout)
model.eval()

total_loss = 0

for inputs, labels in zip(X_test, Y_test):
    # Forward pass
    outputs = model(inputs)

    # Calculate loss (if needed)
    loss = criterion(outputs, labels)
    total_loss += loss.item()

# Calculate average loss
average_loss = total_loss / X_test.shape[0]

# Calculate average evaluation metrics (if needed)
# ...

# Print or store the evaluation results
print(f"Average Loss: {average_loss}")


Average Loss: 0.22212223124600483


## Conclusion

### It's good, very fast framework